### ISSUE TITLE
Graph visualization of subclassed model/layer

### ISSUE THREAD

**Issue by DavidKWH** [tensorflow/tensorflow#31647]
**System information**
- TensorFlow version (you are using): tensorflow-2.0b1
- Are you willing to contribute it (Yes/No): Yes 

**Describe the feature and the current behavior/state.**
Graph visualization of subclassed model/layer/tf.functions

**Will this change the current api? How?**
Not sure

**Who will benefit with this feature?**
All users

**Details**
I posted this question on stack overflow a week ago and it seems it is completely ignored.  This is an attempt to reach out to experts who are more familiar to the inner working of the graph network implementation, and/or tf.function visualization in Tensorboard.  Please forgive me if this is not the right place, and I hope someone here can point me to the right place, thanks.
  
`https://stackoverflow.com/questions/57421766/plotting-subclassed-model-in-tensorflow-2-0-beta`

I have a subclassed model that instantiates a few custom layers via subclassing.  I tried using `keras.utils.plot_model()` but all it does is print the model block, none of the layers appeared.  

Can a Tensorflow expert comment on this?  Will this feature ever be implemented in the future?  If not, what is the next best alternative to examine the computation graph?  Note that `model.summary()` only gives a summary of the parameters of the custom layer, within which contains two dense layers.  Ideally, I like to see all the computations, if that is not asking too much...

**Update**: I dug into the source, looks like plot_model() first check for the `_is_graph_network` attribute.  Graph Networks are used in Functional and Sequential APIs.  From the source:

> Two types of `Networks` exist: Graph Networks and Subclass Networks. Graph
>  networks are used in the Keras Functional and Sequential APIs. Subclassed
>  networks are used when a user subclasses the `Model` class. In general,
>  more Keras features are supported with Graph Networks than with Subclassed
>  Networks, specifically:

>  - Model cloning (`keras.models.clone()`)
>  - Serialization (`model.get_config()/from_config()`, `model.to_json()/to_yaml()`)
>  - Whole-model saving (`model.save()`)

(**custom graph component**)
Naturally, I like to know if I can build a graph network component, so my subclassed model/layer can work with these features.  Does that involve a lot of effort?

(**tf.function graph visualization**)
Can someone let me know if graph visualization via Tensorboard works with Tensorflow2 tf.functions?  In Tensorflow 1.x, one defines a name scope for a logical group of ops (e.g. generator/discriminator in GAN, encoder/decoder in VAE and loss/metrics), they are then displayed as a high-level block in the graph visualization.  Can I define something similar for tf.functions?

**Comment by oanush** [comment]
@gowthamkpr ,
can you please check the issue.Thanks!

**Comment by DavidKWH** [comment]
Thanks, the solution is not applicable to my situation as I am neither
using sequential/functional API, nor calling model.fit().
As stated in my stack overflow question and repeated in the github issue, I
am using a subclassed model and a custom training loop.
In any case, I found most of what I need by reading the source.
Thanks for your help.

**Comment by stephanwlee** [comment]
I do not recall exact details at the moment but I believe there was a innate difficulties with subclassed Model.

@omalleyt12 I vaguely remember that subclassed Layer/Model cannot be guaranteed to be graph friendly and thus fallback to the eager mode. Is this correct? 

Also, reassigning to Tom as he has more expertise on the domain.

**Comment by stephanwlee** [comment]
cc @davidsoergel

**Comment by omalleyt12** [comment]
Yes in general we can't assume anything about the structure of a subclassed Model. If your Model can be though of as blocks of Layers and you wish to visualize it like that, we recommend you view the Functional API

**Comment by innat** [comment]
I've found some workaround to plot with the model sub-classing API. For the obvious reason **Sub-Classing** API doesn't support **Sequential or Functional** API like `model.summary()` and nice visualization using `plot_model`. Here, I will demonstrate both.

```python
class my_model(Model):
    def __init__(self, dim):
        super(my_model, self).__init__()
        self.Base  = VGG16(input_shape=(dim), include_top = False, weights = 'imagenet')
        self.GAP   = L.GlobalAveragePooling2D()
        self.BAT   = L.BatchNormalization()
        self.DROP  = L.Dropout(rate=0.1)
        self.DENS  = L.Dense(256, activation='relu', name = 'dense_A')
        self.OUT   = L.Dense(1, activation='sigmoid')
    
    def call(self, inputs):
        x  = self.Base(inputs)
        g  = self.GAP(x)
        b  = self.BAT(g)
        d  = self.DROP(b)
        d  = self.DENS(d)
        return self.OUT(d)
    
    # AFAIK: The most convenient method to print model.summary() 
    # similar to the sequential or functional API like.
    def build_graph(self):
        x = Input(shape=(dim))
        return Model(inputs=[x], outputs=self.call(x))

dim = (124,124,3)
model = my_model((dim))
model.build((None, *dim))
model.build_graph().summary()
```

It will produce as follows:

```
Layer (type)                 Output Shape              Param #   
=================================================================
input_67 (InputLayer)        [(None, 124, 124, 3)]     0         
_________________________________________________________________
vgg16 (Functional)           (None, 3, 3, 512)         14714688  
_________________________________________________________________
global_average_pooling2d_32  (None, 512)               0         
_________________________________________________________________
batch_normalization_7 (Batch (None, 512)               2048      
_________________________________________________________________
dropout_5 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_A (Dense)              (None, 256)               402192    
_________________________________________________________________
dense_7 (Dense)              (None, 1)                 785       
=================================================================
Total params: 14,848,321
Trainable params: 14,847,297
Non-trainable params: 1,024
```

Now by using the `build_graph` function, we can simply plot the whole architecture. 

```python
# Just showing all possible argument for newcomer.  
tf.keras.utils.plot_model(
    model.build_graph(),                      # here is the trick (for now)
    to_file='model.png', dpi=96,              # saving  
    show_shapes=True, show_layer_names=True,  # show shapes and layer name
    expand_nested=False                       # will show nested block
)
```

It will produce as follows: -)

![a](https://user-images.githubusercontent.com/17668390/93187371-8e545000-f761-11ea-8d70-74dc2fe7c644.png)

**Comment by anugrahasinha** [comment]
I dont think we really need `build_graph` function in the custom model
One can just do this
```
my_model = MyModel()  # You can provide the arguments that 'your model class` requires
# Note : class MyModel(Model), does not have any def build_graph function

inputs = Input(shape=(100,100,3)) # Provide the shape you wish to provide, with respect to 'your model class' expectation
model = Model(inputs=[inputs],outputs=my_model.call(inputs))

model.summary() # This will print the summary properly
tf.keras.utils.plot_model(model,show_shapes=True) # This will give you the visualization of the model.
```

**Comment by innat** [comment]
Yes, we can do that too. But there is nothing big differenc. I mean, here you're doing the same thing without defining it in a function within `Model` class. IMO, defining such function within `Model` class is more intuitive. And also we're building functional api model from subclassed api.

**Comment by prerakmody** [comment]
Following @innat's example, I had to move all my internal functions that I wished to plot to the main models `call()` function. My model is an encoder-decoder architecture ([gist](https://gist.github.com/prerakmody/1b6f349db1634846303fac27060805cc#file-model_subclassing-py))

**Initial Visualization**
![OrganNetPool3](https://user-images.githubusercontent.com/2803936/199478897-7027e461-24cb-4815-9cc0-a749fec01cdd.png)


**After moving submodel `call()` code to main models `call() `**
![OrganNetPool3-v2](https://user-images.githubusercontent.com/2803936/199478921-668c35ae-3e49-4a01-be4b-6ac6b96265ab.png)

I also visualized the .h5 files in netron.app
![image](https://user-images.githubusercontent.com/2803936/199479251-429fddf9-aa43-49f2-a538-8ae17948a8e9.png)

![image](https://user-images.githubusercontent.com/2803936/199479296-b480cfa3-7c5e-41b0-ae75-e6fa1b28e319.png)
